pubkey/src/lib.rs

```diff
@@ -158,10 +160,133 @@ impl From<u64> for PubkeyError {
+    }
+}
+
+#[cfg(all(feature = "rand", not(target_os = "solana")))]
+mod hasher {
+    use rand::Rng;
+
+    /// A special Hasher specifically for Pubkeys. Uses random 8 bytes as
+    /// input. Since pubkeys are effectively random, hashing them is pure overhead.
+    /// Use with care, this may be abusable.
+    #[derive(Default)]
+    pub struct PubkeyHasher {
+        offset: usize,
+        state: u64,
+    }
+
+    impl std::hash::Hasher for PubkeyHasher {
+        #[inline]
+        fn finish(&self) -> u64 {
+            self.state
+        }
+        #[inline]
+        fn write(&mut self, bytes: &[u8]) {
+            let chunk: &[u8; 8] = bytes[self.offset..self.offset + 8].try_into().unwrap();
+            self.state = u64::from_ne_bytes(*chunk);
+        }
+    }
+
+    /// A special HasherBuilder specifically for Pubkeys. Uses random 8 byte slice as
+    /// input. Since pubkeys are effectively random, hashing them is pure overhead.
+    /// Use with care, this is not a secure hasher!
+    #[derive(Clone)]
+    pub struct PubkeyHasherBuilder {
+        offset: usize,
+    }
+
+    impl Default for PubkeyHasherBuilder {
+        fn default() -> Self {
+            std::thread_local!(static OFFSET: core::cell::Cell<usize>  = {
+                let mut rng = rand::prelude::thread_rng();
+                core::cell::Cell::new(rng.gen_range(0..32 - 8))
+            });
+
+            let offset = OFFSET.with(|offset| {
+                let mut next_offset = offset.get() + 1;
+                if next_offset > 24 {
+                    next_offset = 0;
+                }
+                offset.set(next_offset);
+                next_offset
+            });
+            PubkeyHasherBuilder { offset }
+        }
+    }
+
+    impl std::hash::BuildHasher for PubkeyHasherBuilder {
+        type Hasher = PubkeyHasher;
+        #[inline]
+        fn build_hasher(&self) -> Self::Hasher {
+            PubkeyHasher {
+                offset: self.offset,
+                state: 0,
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use {
+            super::PubkeyHasherBuilder,
+            crate::Pubkey,
+            core::hash::{BuildHasher, Hasher},
+        };
+        #[test]
+        fn test_pubkey_hasher_builder() {
+            let key = Pubkey::new_unique();
+            let builder = PubkeyHasherBuilder::default();
+            let mut hasher1 = builder.build_hasher();
+            let mut hasher2 = builder.build_hasher();
+            hasher1.write(key.as_array());
+            hasher2.write(key.as_array());
+            assert_eq!(
+                hasher1.finish(),
+                hasher2.finish(),
+                "Hashers made with same builder should be identical"
+            );
+            // Make sure that when we make new builders we get different slices
+            // chosen for hashing
+            let builder2 = PubkeyHasherBuilder::default();
+            for _ in 0..64 {
```

Comment:
Sorry, I don't understand why this loop is needed. The builder is created outside of the loop, so every hasher from `build_hasher` will use the same offset, and hash the same key.

Was the intention to make the builder inside the loop to use a different offset and check 23 iterations before we get to the original offset from `builder`?

Reply:
Ah gotcha, thanks for the explanation!